sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/CosmosEncryptionAsyncClient.java

```diff
@@ -96,31 +100,45 @@ public EncryptionKeyWrapProvider getEncryptionKeyWrapProvider() {
         String clientEncryptionKeyId,
         String databaseRid,
         CosmosAsyncContainer cosmosAsyncContainer,
-        boolean shouldForceRefresh) {
+        boolean shouldForceRefresh,
+        String ifNoneMatchEtag,
+        boolean shouldForceRefreshGateway) {
         /// Client Encryption key Id is unique within a Database.
         String cacheKey = databaseRid + "/" + clientEncryptionKeyId;
-        if (!shouldForceRefresh) {
+
+        // this allows us to read from the Gateway Cache. If an IfNoneMatchEtag is passed the logic around the gateway
+        // cache allows us to fetch the latest ClientEncryptionKeyProperties from the servers if the gateway cache has
+        // a stale value. This can happen if a client connected via different Gateway has re wrapped the key.
+        RequestOptions requestOptions = new RequestOptions();
+        requestOptions.setHeader(Constants.ALLOW_CACHED_READS_HEADER, String.valueOf(true));
+        requestOptions.setHeader(Constants.DATABASE_RID_HEADER, databaseRid);
+
+        if (StringUtils.isNotEmpty(ifNoneMatchEtag)) {
+            requestOptions.setIfNoneMatchETag(ifNoneMatchEtag);
+        }
+
+        if (!shouldForceRefresh && shouldForceRefreshGateway) {
```

Comment:
`!shouldForceRefresh && shouldForceRefreshGateway  -> !shouldForceRefresh && !shouldForceRefreshGateway`
We want to go gateway when shouldForceRefreshGateway is true ?

Reply:
Thanks for catching this. Fixed

Reply:
There should a test to validate the behaviors when we are refreshing the cache , u can refer EncryptionProcessorAndSettingsTest . Otherwise it would be risky if there is any bug , as all operations will result in meta data call

Reply:
Added the test case